### PR TITLE
requirements met plus first stretch goal, no tests

### PR DESCRIPTION
**Pull request by Bl41r**

### CODE REVIEW

File: flight_paths.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import json
+from simple_graph import Node, SimpleGraph
+from simple_graph import spt_Dijkstra
+
+
+def extract_data_from_json(file):
```
Comment (nhuntwalker) [inline flight_paths.py:6]: Doc string

File: flight_paths.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    for d in data:
+        entry = {}
+        entry['city'] = d['city']
+        entry['neighbors'] = d['destination_cities']
+        entry['lat-lon'] = d['lat_lon']
+        new_data.append(entry)
+    return new_data
+
+
+def build_graph(new_data):
```
Comment (nhuntwalker) [inline flight_paths.py:31]: Doc string.

File: flight_paths.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+def extract_data_from_json(file):
+    with open(file) as json_data:
+        data = json.load(json_data)
+    new_data = []
+    for d in data:
+        entry = {}
+        entry['city'] = d['city']
+        entry['neighbors'] = d['destination_cities']
+        entry['lat-lon'] = d['lat_lon']
+        new_data.append(entry)
```
Comment (nhuntwalker) [inline flight_paths.py:16]: Nice job transforming the data to only take the fields you'd _need_ to use.

File: flight_paths.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    return new_data
+
+
+def build_graph(new_data):
+    gr = SimpleGraph()
```
Comment (nhuntwalker) [inline flight_paths.py:33]: This right here is pretty much **exactly** what I was hoping you'd do.

File: flight_paths.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    for city in new_data:
+        n = Node(city['city'], city['lat-lon'])
+        try:
+            gr.add_node(n)
+        except KeyError:
```
Comment (nhuntwalker) [inline flight_paths.py:39]: When would you get a `KeyError`?

File: flight_paths.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            pass
+
+    for city in new_data:
+        if city['city'] not in gr.node_dict.keys():
+            continue
+        key = city['city']
+        for neighbor in city['neighbors']:
```
Comment (nhuntwalker) [inline flight_paths.py:46]: It's interesting that you'd add neighbors here and not in the previous loop, but I suppose you wanted to build your neighbors list with the distances between cities already baked in.

File: README.md [hunk 7]
```diff
@@ -1,2 +1,15 @@
...
+
+This one features a node search for airplane travel.
+
+https://codefellows.github.io/sea-python-401d4/assignments/kata_flight_paths.html
+
+### usage  in iPython3:
+- run flight_paths.py
+- data = extract_nodes_from_json('cities_with_airports.json')
+- gr = build_graph(data)
+- depth_first_search(gr, gr.node_dict['Denver'], gr.node_dict['Honolulu'])
```
Comment (nhuntwalker) [inline README.md:14]: this is code you should probably include in a tests file!

File: README.md [hunk 8]
```diff
@@ -1,2 +1,15 @@
...
+- spt_Dijkstra(gr, 'Denver', 'Honolulu') OR spt_AStar(gr, 'Denver', 'Honolulu')
```
Comment (nhuntwalker) [inline README.md:15]: Ok Mr. Superstar, doing both Dijkstra's and A-star!

File: README.md [hunk 9]
```diff
@@ -1,2 +1,15 @@
...
+which I have worked on.
+
+
+This one features a node search for airplane travel.
+
+https://codefellows.github.io/sea-python-401d4/assignments/kata_flight_paths.html
+
+### usage  in iPython3:
+- run flight_paths.py
+- data = extract_nodes_from_json('cities_with_airports.json')
```
Comment (nhuntwalker) [inline README.md]: the function name is `extract_data_from_json`, not `nodes`

File: flight_paths.py (hunk 5 above)
Comment (nhuntwalker) [inline flight_paths.py:39]: Getting an unhandled `TypeError` when I try to run your code.

Comment (Bl41r) [reply flight_paths.py:39]: weird, not getting it here, what function?

File: flight_paths.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    radius_earth = 6.371E3 # km
+    phi1 = convert_to_radians(point1[0])
+    phi2 = convert_to_radians(point2[0])
+
+    delta_phi = convert_to_radians(point1[0] - point2[0])
+    delta_lam = convert_to_radians(point1[1] - point2[1])
+
+    a = math.sin(0.5 * delta_phi)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(0.5 * delta_lam)**2
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+    return radius_earth * c / 1.60934   # convert km to miles
```
Comment (nhuntwalker) [inline flight_paths.py:102]: Consider adding the following at the bottom of this file:

``` python
if __name__ == "__main__":
    data = extract_data_from_json('cities_with_airports.json')
    gr = build_graph(data)
    print(depth_first_search(gr, gr.node_dict['Denver'], gr.node_dict['Honolulu']))
```

File: README.md (hunk 9 above)
Comment (Bl41r) [reply README.md]: ah, i changed it last minute but not my copy/paste run code

File: flight_paths.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        entry['lat-lon'] = d['lat_lon']
+        new_data.append(entry)
+    return new_data
+
+
+def build_graph(new_data):
+    gr = SimpleGraph()
+
+    for city in new_data:
+        n = Node(city['city'], city['lat-lon'])
```
Comment (nhuntwalker) [inline flight_paths.py:36]: Your code needs to be python2/3 compatible, as does all code written in this course. This throws a `TypeError` when run in Python 2

File: README.md (hunk 7 above)
Comment (nhuntwalker) [inline README.md:14]: Also, WOW the depth-first path for this connection is ridiculous

File: README.md [hunk 12]
```diff
@@ -1,2 +1,15 @@
...
+This one features a node search for airplane travel.
+
+https://codefellows.github.io/sea-python-401d4/assignments/kata_flight_paths.html
+
+### usage  in iPython3:
+- run flight_paths.py
+- data = extract_data_from_json('cities_with_airports.json')
+- gr = build_graph(data)
+- depth_first_search(gr, gr.node_dict['Denver'], gr.node_dict['Honolulu'])
+- spt_Dijkstra(gr, 'Denver', 'Honolulu') OR spt_AStar(gr, 'Denver', 'Honolulu')
```
Comment (nhuntwalker) [inline README.md:15]: While I appreciate wanting to use what you've already written and not have to rewrite things, it's worth noting that you have print statements in Dijkstra's and A-Star that make the output for each extremely verbose. I'm willing to bet that Will hasn't had a chance to get to your submission for your shortest path assignment yet, but for THIS assignment I just want the output to be the shortest path between the two points and the total distance of that path. Right now I get the shortest distance, which is nice.

File: flight_paths.py (hunk 5 above)
Comment (nhuntwalker) [reply flight_paths.py:39]: It's when run with Python 2. You might want to consider turning everything into unicode literals for this assignment. I think it's `from __future__ import unicode_literals`. I forget right now. WHY ARE YOU AWAKE?!

File: flight_paths.py (hunk 11 above)
Comment (Bl41r) [reply flight_paths.py:36]: fixed

File: flight_paths.py (hunk 10 above)
Comment (Bl41r) [reply flight_paths.py:102]: added!

Comment (nhuntwalker) [reply flight_paths.py:102]: WHY ARE YOU AWAKE!?!

File: README.md (hunk 12 above)
Comment (Bl41r) [reply README.md:15]: ah had debugging code in still.  the Dijkstra and AStar methods I have only get the distance right now, I don't have them to give the short path also.  Right now, just the search will give a path.

File: flight_paths.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    a = math.sin(0.5 * delta_phi)**2 + math.cos(phi1) * math.cos(phi2) * math.sin(0.5 * delta_lam)**2
+    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
+    return radius_earth * c / 1.60934   # convert km to miles
+
+if __name__ == "__main__":
+    data = extract_data_from_json('cities_with_airports.json')
+    gr = build_graph(data)
+    print(depth_first_search(gr, gr.node_dict['Denver'], gr.node_dict['Honolulu']))
+    print('Dijkstra distance:', spt_Dijkstra(gr, 'Denver', 'Honolulu'))
```
Comment (nhuntwalker) [inline flight_paths.py:108]: Your Djikstra's and A\* implementations only return distance. The spec for the stretch goal asked for the path itself as well.

File: test_flight_paths.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+def test_extract_data_from_json():
+    """Test extracted data."""
+    from flight_paths import extract_data_from_json
+    data = extract_data_from_json('cities_with_airports.json')
+    assert len(data) > 0
+    assert type(data) is list
+    assert type(data[0]) is dict
+
+
+def test_build_graph():
```
Comment (nhuntwalker) [inline test_flight_paths.py:12]: This is not a test you should really need. Your graph is coming in more or less unmodified from your data structures, no? And it's tested there, right?

File: flight_paths.py (hunk 5 above)
Comment (nhuntwalker) [reply flight_paths.py:39]: This line (line 39) is untested. Again, when would you get a `KeyError`?

File: test_flight_paths.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+def test_depth_first_search():
+    """Assert search was completed."""
+    from flight_paths import build_graph
+    from flight_paths import extract_data_from_json
+    from flight_paths import depth_first_search
+    data = extract_data_from_json('cities_with_airports.json')
+    gr = build_graph(data)
+    x = depth_first_search(gr, gr.node_dict['Charlotte'], gr.node_dict['Honolulu'])
+    assert x[0][0] == 'Charlotte' and x[0][-1] == 'Honolulu'
```
Comment (nhuntwalker) [inline test_flight_paths.py:33]: What about a test for when there is no path between two cities?
What about a test that when there is a path between two cities with your depth-first search, definite connections are hit along the way?

**Pull request merged**